## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [344]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [345]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [346]:
# Use the describe function to gather some basic statistics
stats = df.describe()
print(stats)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [347]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [348]:
# What three item categories had the most entries?
catagories_counts = df['category'].value_counts()
print(catagories_counts.head(3))







category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [349]:
# For the category with the most entries, which subcategory had the most entries?
subcategory_counts = df.loc[df["category"]== "consumables", "subcategory"].value_counts()
print(subcategory_counts.head(1))


subcategory
bathroom supplies    6424
Name: count, dtype: int64


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [350]:
# Which five clients had the most entries in the data?
top_5_clients = df['client_id'].value_counts().head(5)
top_5_clients




client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [351]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_clients.index.tolist()
top_5_client_ids

[33615, 66037, 46820, 38378, 24741]

[33615, 66037, 46820, 38378, 24741]

In [352]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df["client_id"] == 33615, "qty"].sum()



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [353]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_sumtotal'] = df['unit_price'] * df['qty']
df.head()
#df[['category',	'subcategory',	'unit_price',	'unit_cost', 'unit_weight',	'qty', 'line_number','line_sumtotal']]

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_sumtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [354]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']


def calculate_shipping_price(weight):
    if weight > 50:
        return weight * 7
    else:
        return weight * 10
    


df['Shipping Price'] = df['total_weight'].apply(calculate_shipping_price)

df[['unit_weight', 'qty', 'total_weight', 'unit_price', 'Shipping Price']].head()


unit_weight  qty  total_weight  unit_price  Shipping Price
0         7.50  105        787.50     1096.80         5512.50
1         1.49   21         31.29       24.95          312.90
2         1.68   39         65.52       13.52          458.64
3         1.23   29         35.67       36.42          356.70
4        46.43   20        928.60      195.10         6500.20

In [355]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 1.0925

df['total_price'] = round( (df['line_sumtotal'] + df['Shipping Price']) * sales_tax_rate, 2)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_sumtotal  total_weight  Shipping Price  \
0         7.50  105            1      115164.00        787.50         5512.50   
1         1.49   21            0         523.95         31.29          312.90   
2         1.68   39            6         527.28         65.52          458.64   
3         1.23   29            3        1056.18         35.67          356.70   
4        46.43   20            1        3902.00        928.60         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 22 columns]

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>line_subtotal</th>
      <th>shipping_price</th>
      <th>line_price</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>115164.00</td>
      <td>5512.50</td>
      <td>131839.08</td>
    </tr>
    <tr>
      <th>1</th>
      <td>523.95</td>
      <td>312.90</td>
      <td>914.26</td>
    </tr>
    <tr>
      <th>2</th>
      <td>527.28</td>
      <td>458.64</td>
      <td>1077.12</td>
    </tr>
  </tbody>
</table>
</div>

In [356]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] +  df['Shipping Price']

df




first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_sumtotal  total_weight  Shipping Price  total_price  \
0                1      115164.00        787.50         5512.50    131839.08   
1                0         523.95         31.29          312.90       914.26   
2                6         527.28         65.52          458.64      1077.12   
3                3        1056.18         35.67          356.70      1543.57   
4                1        3902.00        928.60         6500.20     11364.40   
...            ...            ...           ...             ...          ...   
54634            8        2743.29         74.25          519.75      3564.87   
54635            0        9709.73        549.90         3849.30     14813.24   
54636            6       31188.50       1976.00        13832.00     49184.90   
54637            0         165.76       2020.48        14143.36     15632.71   
54638            5        3103.31       2134.17        14939.19     19711.43   

           cost  
0      85597.05  
1        629.79  
2        765.18  
3       1077.35  
4       8663.60  
...         ...  
54634   2222.55  
54635  12095.92  
54636  41054.25  
54637  14280.00  
54638  16599.10  

[54639 rows x 23 columns]

In [357]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - (df['cost'])
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       line_sumtotal  total_weight  Shipping Price  total_price      cost  \
0          115164.00        787.50         5512.50    131839.08  85597.05   
1             523.95         31.29          312.90       914.26    629.79   
2             527.28         65.52          458.64      1077.12    765.18   
3            1056.18         35.67          356.70      1543.57   1077.35   
4            3902.00        928.60         6500.20     11364.40   8663.60   
...              ...           ...             ...          ...       ...   
54634        2743.29         74.25          519.75      3564.87   2222.55   
54635        9709.73        549.90         3849.30     14813.24  12095.92   
54636       31188.50       1976.00        13832.00     49184.90  41054.25   
54637         165.76       2020.48        14143.36     15632.71  14280.00   
54638        3103.31       2134.17        14939.19     19711.43  16599.10   

         profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [358]:
# Check your work using the totals above
order_id_list = [2742071, 2173913, 6128929] 
def sum_of_price (order_id):
    order_line = df.loc[df['order_id'] == order_id, 'total_price']
    return round (order_line.sum(), 2)
for order_id in  order_id_list: 
    order_price = sum_of_price(order_id)
    print(f"Order ID: {order_id}, Total Price: {order_price}")





    


Order ID: 2742071, Total Price: 152811.89
Order ID: 2173913, Total Price: 162388.71
Order ID: 6128929, Total Price: 923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [359]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_clients = [33615, 66037, 46820, 38378, 24741]

df




first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       line_sumtotal  total_weight  Shipping Price  total_price      cost  \
0          115164.00        787.50         5512.50    131839.08  85597.05   
1             523.95         31.29          312.90       914.26    629.79   
2             527.28         65.52          458.64      1077.12    765.18   
3            1056.18         35.67          356.70      1543.57   1077.35   
4            3902.00        928.60         6500.20     11364.40   8663.60   
...              ...           ...             ...          ...       ...   
54634        2743.29         74.25          519.75      3564.87   2222.55   
54635        9709.73        549.90         3849.30     14813.24  12095.92   
54636       31188.50       1976.00        13832.00     49184.90  41054.25   
54637         165.76       2020.48        14143.36     15632.71  14280.00   
54638        3103.31       2134.17        14939.19     19711.43  16599.10   

         profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 24 columns]

In [360]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
summary_df = df.groupby('client_id').agg({
    'qty': 'sum',
    'Shipping Price': 'sum',
    'total_price': 'sum',
    'profit': 'sum'
}).reset_index()

# Rename the columns for clarity
summary_df = summary_df.rename(columns={
    'qty': 'total_units_purchased',
    'Shipping Price': 'total_shipping_price',
    'total_price': 'total_revenue',
    'profit': 'total_profit'
})

# Sort the data by total revenue in descending order and select the top 5 clients
summary_df = summary_df.sort_values(by='total_revenue', ascending=False).head(5)

print(summary_df)


     client_id  total_units_purchased  total_shipping_price  total_revenue  \
758      78965                4033302            5340756.88   6.559237e+08   
840      85931                 644423           10122549.19   2.374170e+08   
320      38473                 402770            4391003.73   1.835526e+08   
850      86721                1010426           64335727.72   1.812663e+08   
821      84344                 466968            4746583.13   1.066481e+08   

     total_profit  
758  2.473528e+08  
840  1.092941e+08  
320  4.335173e+07  
850  4.297470e+07  
821  2.771170e+07  


In [361]:
# Format the data and rename the columns to names suitable for presentation.
df.style \
    .format({'total_revenue': '{:,.2f}', 'total_profit': '{:,.2f}'}) \
    .background_gradient(cmap='viridis', subset=['total_revenue', 'total_profit']) \
    .bar(subset=['total_units_purchased'], color='#d65f5f')

KeyError: "None of [Index(['total_revenue', 'total_profit'], dtype='object')] are in the [columns]"

In [362]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

sorted_data = data.sort_values(by='total_profit (millions)', ascending=False)
print(f"Top {len(sorted_data)} rows of data sorted by Total Profit (millions) in descending order:")
print(sorted_data.head())


KeyError: 'total_profit (millions)'